### PR TITLE
Remove deprecated LLVM-style inline assembly

### LINKED ISSUE #70173: Tracking Issue for LLVM-style inline assembly (`llvm_asm`)
This is a tracking issue for the RFC "2843" (rust-lang/rfcs#2843).
The feature gate for the issue is `#![feature(llvm_asm)]`.

This issue tracks the legacy inline assembly macro `llvm_asm!` which more-or-less maps directly to LLVM's internal representation of inline assembly.

There is currently no plan for stabilizing this macro, but there is a proposal to replace it with the new inline assembly macro specified in RFC 2873 (rust-lang/rfcs#2873).

Implementation PR: #68404

### LINKED ISSUE #82065: SIGSEGV in the compiler
### Code

```Rust
#![feature(llvm_asm)]
fn main() {
    unsafe { llvm_asm!("mov $0, %rax" :: "i"(main)); }
}
```

### Meta

`rustc --version --verbose`:
```
rustc 1.50.0-nightly (f74583445 2020-12-18)
binary: rustc
commit-hash: f74583445702e2e27ec4415376f2c540a83d7ded
commit-date: 2020-12-18
host: x86_64-unknown-linux-gnu
release: 1.50.0-nightly
```

`rustc +stable --version --verbose`:
```
rustc 1.50.0 (cb75ad5db 2021-02-10)
binary: rustc
commit-hash: cb75ad5db02783e8b0222fee363c5f63f7e2cf5b
commit-date: 2021-02-10
host: x86_64-unknown-linux-gnu
release: 1.50.0
```
Note that on stable `RUSTC_BOOTSTRAP=1` is needed to trigger the bug

### Error output
Hmm. What should go here?
```
Segmentation fault (core dumped)
```

<details><summary><strong>GDB Backtrace</strong></summary>
<p>

```
0x00007ffff29005a9 in llvm::X86TargetLowering::LowerAsmOperandForConstraint(llvm::SDValue, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&, std::vector<llvm::SDValue, std::allocator<llvm::SDValue> >&, llvm::SelectionDAG&) const () from /home/pitust/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/bin/../lib/../lib/libLLVM-11-rust-1.50.0-nightly.so
(gdb) bt
#0  0x00007ffff29005a9 in llvm::X86TargetLowering::LowerAsmOperandForConstraint(llvm::SDValue, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&, std::vector<llvm::SDValue, std::allocator<llvm::SDValue> >&, llvm::SelectionDAG&) const () from /home/pitust/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/bin/../lib/../lib/libLLVM-11-rust-1.50.0-nightly.so
#1  0x00007ffff0a2241a in llvm::SelectionDAGBuilder::visitInlineAsm(llvm::CallBase const&) ()
   from /home/pitust/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/bin/../lib/../lib/libLLVM-11-rust-1.50.0-nightly.so
#2  0x00007ffff09ff867 in llvm::SelectionDAGBuilder::visit(llvm::Instruction const&) () from /home/pitust/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/bin/../lib/../lib/libLLVM-11-rust-1.50.0-nightly.so
#3  0x00007ffff0abf4fe in llvm::SelectionDAGISel::SelectBasicBlock(llvm::ilist_iterator<llvm::ilist_detail::node_options<llvm::Instruction, false, false, void>, false, true>, llvm::ilist_iterator<llvm::ilist_detail::node_options<llvm::Instruction, false, false, void>, false, true>, bool&) () from /home/pitust/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/bin/../lib/../lib/libLLVM-11-rust-1.50.0-nightly.so
#4  0x00007ffff0abe793 in llvm::SelectionDAGISel::SelectAllBasicBlocks(llvm::Function const&) ()
   from /home/pitust/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/bin/../lib/../lib/libLLVM-11-rust-1.50.0-nightly.so
#5  0x00007ffff0abb926 in llvm::SelectionDAGISel::runOnMachineFunction(llvm::MachineFunction&) ()
   from /home/pitust/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/bin/../lib/../lib/libLLVM-11-rust-1.50.0-nightly.so
#6  0x00007ffff27dacd7 in (anonymous namespace)::X86DAGToDAGISel::runOnMachineFunction(llvm::MachineFunction&) ()
--Type <RET> for more, q to quit, c to continue without paging--
   from /home/pitust/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/bin/../lib/../lib/libLLVM-11-rust-1.50.0-nightly.so
#7  0x00007ffff062478e in llvm::MachineFunctionPass::runOnFunction(llvm::Function&) () from /home/pitust/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/bin/../lib/../lib/libLLVM-11-rust-1.50.0-nightly.so
#8  0x00007ffff03fefd2 in llvm::FPPassManager::runOnFunction(llvm::Function&) () from /home/pitust/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/bin/../lib/../lib/libLLVM-11-rust-1.50.0-nightly.so
#9  0x00007ffff04059e3 in llvm::FPPassManager::runOnModule(llvm::Module&) () from /home/pitust/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/bin/../lib/../lib/libLLVM-11-rust-1.50.0-nightly.so
#10 0x00007ffff03ffa2a in llvm::legacy::PassManagerImpl::run(llvm::Module&) () from /home/pitust/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/bin/../lib/../lib/libLLVM-11-rust-1.50.0-nightly.so
#11 0x00007ffff40240ab in LLVMRustWriteOutputFile () from /home/pitust/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/bin/../lib/librustc_driver-54c3daf069bcd3f6.so
#12 0x00007ffff3f2656a in rustc_codegen_llvm::back::write::write_output_file () from /home/pitust/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/bin/../lib/librustc_driver-54c3daf069bcd3f6.so
#13 0x00007ffff3f2b604 in rustc_codegen_llvm::back::write::codegen () from /home/pitust/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/bin/../lib/librustc_driver-54c3daf069bcd3f6.so
#14 0x00007ffff3f4997b in rustc_codegen_ssa::back::write::finish_intra_module_work () from /home/pitust/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/bin/../lib/librustc_driver-54c3daf069bcd3f6.so
#15 0x00007ffff3f43cd7 in rustc_codegen_ssa::back::write::execute_work_item () from /home/pitust/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/bin/../lib/librustc_driver-54c3daf069bcd3f6.so
#16 0x00007ffff3e64d3a in std::sys_common::backtrace::__rust_begin_short_backtrace () from /home/pitust/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/bin/../lib/librustc_driver-54c3daf069bcd3f6.so
#17 0x00007ffff3e6ac6d in core::ops::function::FnOnce::call_once{{vtable-shim}} () from /home/pitust/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/bin/../lib/librustc_driver-54c3daf069bcd3f6.so
--Type <RET> for more, q to quit, c to continue without paging--
#18 0x00007ffff32717ba in alloc::boxed::{{impl}}::call_once<(),FnOnce<()>,alloc::alloc::Global> () at /rustc/f74583445702e2e27ec4415376f2c540a83d7ded/library/alloc/src/boxed.rs:1328
#19 alloc::boxed::{{impl}}::call_once<(),alloc::boxed::Box<FnOnce<()>, alloc::alloc::Global>,alloc::alloc::Global> () at /rustc/f74583445702e2e27ec4415376f2c540a83d7ded/library/alloc/src/boxed.rs:1328
#20 std::sys::unix::thread::{{impl}}::new::thread_start () at /rustc/f74583445702e2e27ec4415376f2c540a83d7ded//library/std/src/sys/unix/thread.rs:71
#21 0x00007ffff317a299 in start_thread () from /usr/lib/libpthread.so.0
#22 0x00007ffff3091153 in clone () from /usr/lib/libc.so.6
```

</p>
</details>

I have no clue why this happens, but i guess writing a JIT in rust like this wasn't the smartest idea...

### LINKED ISSUE #87612: Segfault when using llvm_asm memory constraint
I tried this code:

```rust
#![feature(llvm_asm)]
#![crate_type="staticlib"]

// using no_mangle to force codegen, otherwise nothing is emitted
#[no_mangle]
pub unsafe fn test() -> u32 {
    let test: u32;
    llvm_asm!("mov $$32, $0" : "=m"(test));
    test
}
```

I expected to see this happen:
```
> rustc test.rs
<no output>
```

Instead, this happened:
```
> rustc test.rs 
/home/tuomas/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/bin/../lib/librustc_driver-41fdbb8b89e55544.so(+0x4fdb63)[0x7f6a182c2b63]
/lib/x86_64-linux-gnu/libpthread.so.0(+0x153c0)[0x7f6a17a1f3c0]
/home/tuomas/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/bin/../lib/../lib/libLLVM-12-rust-1.56.0-nightly.so(_ZN4llvm19SelectionDAGBuilder14visitInlineAsmERKNS_8CallBaseE+0x1621)[0x7f6a14f71c31]
/home/tuomas/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/bin/../lib/../lib/libLLVM-12-rust-1.56.0-nightly.so(_ZN4llvm19SelectionDAGBuilder5visitERKNS_11InstructionE+0x67)[0x7f6a14f4ff07]
/home/tuomas/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/bin/../lib/../lib/libLLVM-12-rust-1.56.0-nightly.so(_ZN4llvm16SelectionDAGISel16SelectBasicBlockENS_14ilist_iteratorINS_12ilist_detail12node_optionsINS_11InstructionELb0ELb0EvEELb0ELb1EEES6_Rb+0x12e)[0x7f6a1500f3be]
/home/tuomas/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/bin/../lib/../lib/libLLVM-12-rust-1.56.0-nightly.so(_ZN4llvm16SelectionDAGISel20SelectAllBasicBlocksERKNS_8FunctionE+0x1a3b)[0x7f6a1500e49b]
/home/tuomas/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/bin/../lib/../lib/libLLVM-12-rust-1.56.0-nightly.so(_ZN4llvm16SelectionDAGISel20runOnMachineFunctionERNS_15MachineFunctionE+0xb1d)[0x7f6a1500ba9d]
/home/tuomas/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/bin/../lib/../lib/libLLVM-12-rust-1.56.0-nightly.so(+0x48c1177)[0x7f6a16f13177]
/home/tuomas/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/bin/../lib/../lib/libLLVM-12-rust-1.56.0-nightly.so(_ZN4llvm19MachineFunctionPass13runOnFunctionERNS_8FunctionE+0xee)[0x7f6a14b3496e]
/home/tuomas/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/bin/../lib/../lib/libLLVM-12-rust-1.56.0-nightly.so(_ZN4llvm13FPPassManager13runOnFunctionERNS_8FunctionE+0x639)[0x7f6a148f7869]
/home/tuomas/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/bin/../lib/../lib/libLLVM-12-rust-1.56.0-nightly.so(_ZN4llvm13FPPassManager11runOnModuleERNS_6ModuleE+0x33)[0x7f6a148fe703]
/home/tuomas/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/bin/../lib/../lib/libLLVM-12-rust-1.56.0-nightly.so(_ZN4llvm6legacy15PassManagerImpl3runERNS_6ModuleE+0x470)[0x7f6a148f80d0]
/home/tuomas/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/bin/../lib/librustc_driver-41fdbb8b89e55544.so(+0x20c41e8)[0x7f6a19e891e8]
/home/tuomas/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/bin/../lib/librustc_driver-41fdbb8b89e55544.so(+0x2050c3f)[0x7f6a19e15c3f]
/home/tuomas/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/bin/../lib/librustc_driver-41fdbb8b89e55544.so(+0x2053afc)[0x7f6a19e18afc]
/home/tuomas/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/bin/../lib/librustc_driver-41fdbb8b89e55544.so(+0x204f29b)[0x7f6a19e1429b]
/home/tuomas/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/bin/../lib/librustc_driver-41fdbb8b89e55544.so(+0x2048ada)[0x7f6a19e0dada]
/home/tuomas/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/bin/../lib/librustc_driver-41fdbb8b89e55544.so(+0x20940bc)[0x7f6a19e590bc]
/home/tuomas/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/bin/../lib/librustc_driver-41fdbb8b89e55544.so(+0x20b267c)[0x7f6a19e7767c]
/home/tuomas/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/bin/../lib/libstd-9c44f1fe63de3edb.so(rust_metadata_std_ed022288c3a1d27e+0xa8a97)[0x7f6a17aeda97]
/lib/x86_64-linux-gnu/libpthread.so.0(+0x9609)[0x7f6a17a13609]
/lib/x86_64-linux-gnu/libc.so.6(clone+0x43)[0x7f6a17927293]
Segmentation fault
```

### Meta

`rustc --version --verbose`:
```
rustc 1.56.0-nightly (492723897 2021-07-29)
binary: rustc
commit-hash: 492723897e9b4db6701b3a75b72618d08a7d5319
commit-date: 2021-07-29
host: x86_64-unknown-linux-gnu
release: 1.56.0-nightly
LLVM version: 12.0.1
```

<details><summary>Backtrace</summary>
<p>

```
> BACKTRACE=1 rustc test.rs
/home/tuomas/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/bin/../lib/librustc_driver-41fdbb8b89e55544.so(+0x4fdb63)[0x7f6d10697b63]
/lib/x86_64-linux-gnu/libpthread.so.0(+0x153c0)[0x7f6d0fdf43c0]
/home/tuomas/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/bin/../lib/../lib/libLLVM-12-rust-1.56.0-nightly.so(_ZN4llvm19SelectionDAGBuilder14visitInlineAsmERKNS_8CallBaseE+0x1621)[0x7f6d0d346c31]
/home/tuomas/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/bin/../lib/../lib/libLLVM-12-rust-1.56.0-nightly.so(_ZN4llvm19SelectionDAGBuilder5visitERKNS_11InstructionE+0x67)[0x7f6d0d324f07]
/home/tuomas/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/bin/../lib/../lib/libLLVM-12-rust-1.56.0-nightly.so(_ZN4llvm16SelectionDAGISel16SelectBasicBlockENS_14ilist_iteratorINS_12ilist_detail12node_optionsINS_11InstructionELb0ELb0EvEELb0ELb1EEES6_Rb+0x12e)[0x7f6d0d3e43be]
/home/tuomas/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/bin/../lib/../lib/libLLVM-12-rust-1.56.0-nightly.so(_ZN4llvm16SelectionDAGISel20SelectAllBasicBlocksERKNS_8FunctionE+0x1a3b)[0x7f6d0d3e349b]
/home/tuomas/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/bin/../lib/../lib/libLLVM-12-rust-1.56.0-nightly.so(_ZN4llvm16SelectionDAGISel20runOnMachineFunctionERNS_15MachineFunctionE+0xb1d)[0x7f6d0d3e0a9d]
/home/tuomas/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/bin/../lib/../lib/libLLVM-12-rust-1.56.0-nightly.so(+0x48c1177)[0x7f6d0f2e8177]
/home/tuomas/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/bin/../lib/../lib/libLLVM-12-rust-1.56.0-nightly.so(_ZN4llvm19MachineFunctionPass13runOnFunctionERNS_8FunctionE+0xee)[0x7f6d0cf0996e]
/home/tuomas/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/bin/../lib/../lib/libLLVM-12-rust-1.56.0-nightly.so(_ZN4llvm13FPPassManager13runOnFunctionERNS_8FunctionE+0x639)[0x7f6d0cccc869]
/home/tuomas/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/bin/../lib/../lib/libLLVM-12-rust-1.56.0-nightly.so(_ZN4llvm13FPPassManager11runOnModuleERNS_6ModuleE+0x33)[0x7f6d0ccd3703]
/home/tuomas/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/bin/../lib/../lib/libLLVM-12-rust-1.56.0-nightly.so(_ZN4llvm6legacy15PassManagerImpl3runERNS_6ModuleE+0x470)[0x7f6d0cccd0d0]
/home/tuomas/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/bin/../lib/librustc_driver-41fdbb8b89e55544.so(+0x20c41e8)[0x7f6d1225e1e8]
/home/tuomas/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/bin/../lib/librustc_driver-41fdbb8b89e55544.so(+0x2050c3f)[0x7f6d121eac3f]
/home/tuomas/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/bin/../lib/librustc_driver-41fdbb8b89e55544.so(+0x2053afc)[0x7f6d121edafc]
/home/tuomas/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/bin/../lib/librustc_driver-41fdbb8b89e55544.so(+0x204f29b)[0x7f6d121e929b]
/home/tuomas/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/bin/../lib/librustc_driver-41fdbb8b89e55544.so(+0x2048ada)[0x7f6d121e2ada]
/home/tuomas/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/bin/../lib/librustc_driver-41fdbb8b89e55544.so(+0x20940bc)[0x7f6d1222e0bc]
/home/tuomas/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/bin/../lib/librustc_driver-41fdbb8b89e55544.so(+0x20b267c)[0x7f6d1224c67c]
/home/tuomas/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/bin/../lib/libstd-9c44f1fe63de3edb.so(rust_metadata_std_ed022288c3a1d27e+0xa8a97)[0x7f6d0fec2a97]
/lib/x86_64-linux-gnu/libpthread.so.0(+0x9609)[0x7f6d0fde8609]
/lib/x86_64-linux-gnu/libc.so.6(clone+0x43)[0x7f6d0fcfc293]
Segmentation fault
```

</p>
</details>

<details><summary>GDB Backtrace</summary>
<p>

```gdb
Reading symbols from rustc...
(gdb) run test.rs
Starting program: /home/tuomas/.cargo/bin/rustc test.rs
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
process 12241 is executing new program: /home/tuomas/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/bin/rustc
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
[New Thread 0x7fffee3ff700 (LWP 12245)]
[New Thread 0x7fffe8387700 (LWP 12246)]
[New Thread 0x7fffe7dff700 (LWP 12247)]
[New Thread 0x7fffe73ff700 (LWP 12248)]
[New Thread 0x7fffe6dff700 (LWP 12249)]

Thread 6 "opt test.6a6880" received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x7fffe6dff700 (LWP 12249)]
0x00007ffff1809c31 in llvm::SelectionDAGBuilder::visitInlineAsm(llvm::CallBase const&) ()
   from /home/tuomas/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/bin/../lib/../lib/libLLVM-12-rust-1.56.0-nightly.so
(gdb) bt
#0  0x00007ffff1809c31 in llvm::SelectionDAGBuilder::visitInlineAsm(llvm::CallBase const&) ()
   from /home/tuomas/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/bin/../lib/../lib/libLLVM-12-rust-1.56.0-nightly.so
#1  0x00007ffff17e7f07 in llvm::SelectionDAGBuilder::visit(llvm::Instruction const&) ()
   from /home/tuomas/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/bin/../lib/../lib/libLLVM-12-rust-1.56.0-nightly.so
#2  0x00007ffff18a73be in llvm::SelectionDAGISel::SelectBasicBlock(llvm::ilist_iterator<llvm::ilist_detail::node_options<llvm::Instruction, false, false, void>, false, true>, llvm::ilist_iterator<llvm::ilist_detail::node_options<llvm::Instruction, false, false, void>, false, true>, bool&) ()
   from /home/tuomas/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/bin/../lib/../lib/libLLVM-12-rust-1.56.0-nightly.so
#3  0x00007ffff18a649b in llvm::SelectionDAGISel::SelectAllBasicBlocks(llvm::Function const&) ()
   from /home/tuomas/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/bin/../lib/../lib/libLLVM-12-rust-1.56.0-nightly.so
#4  0x00007ffff18a3a9d in llvm::SelectionDAGISel::runOnMachineFunction(llvm::MachineFunction&) ()
   from /home/tuomas/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/bin/../lib/../lib/libLLVM-12-rust-1.56.0-nightly.so
#5  0x00007ffff37ab177 in (anonymous namespace)::X86DAGToDAGISel::runOnMachineFunction(llvm::MachineFunction&) [clone .llvm.6640036156556231831] ()
   from /home/tuomas/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/bin/../lib/../lib/libLLVM-12-rust-1.56.0-nightly.so
#6  0x00007ffff13cc96e in llvm::MachineFunctionPass::runOnFunction(llvm::Function&) ()
   from /home/tuomas/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/bin/../lib/../lib/libLLVM-12-rust-1.56.0-nightly.so
#7  0x00007ffff118f869 in llvm::FPPassManager::runOnFunction(llvm::Function&) ()
   from /home/tuomas/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/bin/../lib/../lib/libLLVM-12-rust-1.56.0-nightly.so
#8  0x00007ffff1196703 in llvm::FPPassManager::runOnModule(llvm::Module&) ()
   from /home/tuomas/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/bin/../lib/../lib/libLLVM-12-rust-1.56.0-nightl--Type <RET> for more, q to quit, c to continue without paging--
y.so
#9  0x00007ffff11900d0 in llvm::legacy::PassManagerImpl::run(llvm::Module&) ()
   from /home/tuomas/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/bin/../lib/../lib/libLLVM-12-rust-1.56.0-nightly.so
#10 0x00007ffff67211e8 in LLVMRustWriteOutputFile ()
   from /home/tuomas/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/bin/../lib/librustc_driver-41fdbb8b89e55544.so
#11 0x00007ffff66adc3f in rustc_codegen_llvm::back::write::write_output_file ()
   from /home/tuomas/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/bin/../lib/librustc_driver-41fdbb8b89e55544.so
#12 0x00007ffff66b0afc in rustc_codegen_llvm::back::write::codegen ()
   from /home/tuomas/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/bin/../lib/librustc_driver-41fdbb8b89e55544.so
#13 0x00007ffff66ac29b in rustc_codegen_ssa::back::write::finish_intra_module_work ()
   from /home/tuomas/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/bin/../lib/librustc_driver-41fdbb8b89e55544.so
#14 0x00007ffff66a5ada in rustc_codegen_ssa::back::write::execute_work_item ()
   from /home/tuomas/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/bin/../lib/librustc_driver-41fdbb8b89e55544.so
#15 0x00007ffff66f10bc in std::sys_common::backtrace::__rust_begin_short_backtrace ()
   from /home/tuomas/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/bin/../lib/librustc_driver-41fdbb8b89e55544.so
#16 0x00007ffff670f67c in core::ops::function::FnOnce::call_once{{vtable-shim}} ()
   from /home/tuomas/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/bin/../lib/librustc_driver-41fdbb8b89e55544.so
#17 0x00007ffff4385a97 in <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once ()
    at /rustc/492723897e9b4db6701b3a75b72618d08a7d5319/library/alloc/src/boxed.rs:1572
#18 <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once ()
    at /rustc/492723897e9b4db6701b3a75b72618d08a7d5319/library/alloc/src/boxed.rs:1572
#19 std::sys::unix::thread::Thread::new::thread_start () at library/std/src/sys/unix/thread.rs:91
#20 0x00007ffff42ab609 in start_thread (arg=<optimized out>) at pthread_create.c:477
#21 0x00007ffff41bf293 in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95
```
</p>
</details>

### LINKED ISSUE #92794: llvm_asm!() with indirects broken on LLVM 14
This broke at LLVM change bec726f5d20ecfdc4a1cd4076ddc6edb35a317e5 - I've spent some time poking at it but it looks like it may require the attention of someone a little more familiar with inline assembly than me. Basically we need to explicitly emit the elementtype(TYPE) attribute for each indirect argument of an inline assembly call. That's mostly easy for outputs, but for inputs I think it's going to take a bit of a refactor to make it work.

### PR DESCRIPTION
**Pull request by tmiasko**
The `llvm_asm!` was deprecated back in #87590 1.56.0, with intention to remove
it once `asm!` was stabilized, which already happened in #91728 1.59.0. Now it
is time to remove `llvm_asm!` to avoid continued maintenance cost.

Closes #70173.
Closes #92794.
Closes #87612.
Closes #82065.

cc @rust-lang/wg-inline-asm

r? @Amanieu

### CODE REVIEW

**Comment by Amanieu** [comment]
`compiler/rustc_builtin_macros/src/asm.rs` has a suggestion to use `llvm_asm!` if it detects the legacy syntax. I think this detection code should just be removed entirely since the renaming of the old `asm!` to `llvm_asm!` happened 2 years ago.

File: compiler/rustc_ast/src/ast.rs [hunk 1]
```diff
@@ -2085,32 +2083,6 @@ pub enum LlvmAsmDialect {
     Intel,
 }
```
[the remark this answers is not in the document]
Comment (tmiasko) [reply compiler/rustc_ast/src/ast.rs]: Thanks. Removed / migrated remaining uses in LLVM codegen to `llvm::AsmDialect`.

**Comment by tmiasko** [comment]
> `compiler/rustc_builtin_macros/src/asm.rs` has a suggestion to use `llvm_asm!` if it detects the legacy syntax. I think this detection code should just be removed entirely since the renaming of the old `asm!` to `llvm_asm!` happened 2 years ago.

Thanks. Removed.

**Comment by bjorn3** [comment]
Those are nice perf improvements!

**Comment by Amanieu** [comment]
I was honestly not expecting that. It seems that MIR borrowck somehow got a speedup, most likely due to the removal of `MutateMode`?

**Comment by tmiasko** [comment]
In keccak-Chec-Full perf difference seems incidental, code generated for a very hot loop in `iterate_fixpoint` improved quite a bit):

```
--------------------------------------------------------------------------------
Ir             
--------------------------------------------------------------------------------
-1,715,117,584  PROGRAM TOTALS

--------------------------------------------------------------------------------
Ir             file:function
--------------------------------------------------------------------------------
 -596,852,099  ???:<rustc_mir_dataflow::framework::engine::Engine<rustc_mir_dataflow::impls::MaybeInitializedPlaces>>::iterate_to_fixpoint
 -596,805,748  ???:<rustc_mir_dataflow::framework::engine::Engine<rustc_mir_dataflow::impls::MaybeUninitializedPlaces>>::iterate_to_fixpoint
 -523,105,586  ???:<rustc_mir_dataflow::framework::engine::Engine<rustc_mir_dataflow::impls::EverInitializedPlaces>>::iterate_to_fixpoint
```

This seems more or less what happened in a different PR (similar diff / codegen change) https://perf.rust-lang.org/compare.html?start=38c22af0153cf8f920c01ef04493e8878401fd18&end=8c4a951b919d548bba16b6f209ed05b1647eba2c.

cc @oli-obk I suspect that in #92928 & #93007 you might be chasing a regression that is no longer there.

**Pull request merged**